### ISSUE TITLE
Wrong response.body encoding with http-equiv headers

### ISSUE THREAD

**Issue by nyov** [scrapy/scrapy#4933]
A `Response` object doesn't seem to obey a `http-equiv` header for `Content-Type` encoding when it found a HTTP header saying different.
So if the http header says 'utf-8' but the body content is, say, codepage 1252 and the documents' http-equiv says 1252, then scrapy appears to still picks utf-8 for decoding body content.

That might be the right decision, but I think it's wrong. The document itself should know it's encoding better than a server-wide setting would.

**Comment by nyov** [comment]
Well, browsers do the same, apparently. Sticking to browser behaviour might be saner, but I still don't like it.
An option to change the behaviour would be nice.
`response.encoding` is a property, but it can't be set. Modifying the response requires building a new response, or am I missing something? The order of encoding detection is hard-coded into `TextResponse` class and going for a custom response class doesn't seem an easy option, either?

Example session. A server that throws blanket `Content-Type: text/html; charset=utf-8` responses, with a document that has a `<meta http-equiv="Content-Type" content="text/html; charset=windows-1252" />` line:
```py
$ scrapy-git shell 'http://127.2.2.2/test.html'
2020-12-20 13:07:11 [scrapy.utils.log] INFO: Scrapy 2.4.1 started (bot: scrapybot)
[...SNIP...]
2020-12-20 13:07:12 [scrapy.core.engine] INFO: Spider opened
2020-12-20 13:07:12 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://127.2.2.2/test.html> (referer: None)
[s] Available Scrapy objects:
[s]   scrapy     scrapy module (contains scrapy.Request, scrapy.Selector, etc)
[s]   crawler    <scrapy.crawler.Crawler object at 0x7fd292aa0e80>
[s]   item       {}
[s]   request    <GET http://127.2.2.2/test.html>
[s]   response   <200 http://127.2.2.2/test.html>
[s]   settings   <scrapy.settings.Settings object at 0x7fd292aa0f98>
[s]   spider     <DefaultSpider 'default' at 0x7fd292796da0>
[s] Useful shortcuts:
[s]   fetch(url[, redirect=True]) Fetch URL and update local objects (by default, redirects are followed)
[s]   fetch(req)                  Fetch a scrapy.Request and update local objects 
[s]   shelp()           Shell help (print this help)
[s]   view(response)    View response in a browser
In [1]: print(response.body.decode(response.encoding))
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-1-547102fac1e0> in <module>()
----> 1 print(response.body.decode(response.encoding))

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe6 in position 166: invalid continuation byte

In [2]: response.encoding
Out[2]: 'utf-8'

In [3]: print(response.body.decode('cp1252'))
<html>
  <head>
  <meta http-equiv="Content-Type" content="text/html; charset=windows-1252" />
  <meta http-equiv="Content-Language" content="en" />
</head>
<body>
	hællö wörlð
	hullo wärld

	how iß ze life?
</body>
</html>


In [4]: response.headers
Out[4]: 
{b'Date': b'Sun, 20 Dec 2020 13:07:12 GMT',
 b'Last-Modified': b'Sun, 20 Dec 2020 13:04:38 GMT',
 b'Etag': b'"e0-5b6e4fbe21086"',
 b'Accept-Ranges': b'bytes',
 b'Content-Type': b'text/html; charset=utf-8'}

In [5]: 
```

**Comment by Gallaecio** [comment]
This could be implemented in a downloader middleware though, right? The middleware could detect these cases and re-create the response accordingly.

**Comment by majate** [comment]
As a university assignment a small group of students (including me) are planning to take on this issue. Our plan is to implement a new downloader middleware that updates the response encoding as requested. This middleware will be added to `DOWNLOADER_MIDDLEWARES_BASE`, but with an enable flag defaulting to false. Does this seem like an appropriate approach?

**Comment by Gallaecio** [comment]
@majate If your end goal is to get your changes eventually merged, this might not be the best issue for you to tackle.

My middleware suggestion was just an idea, it may not be the best approach. For example, if there is a response encoded with something that UTF-8 cannot decode, the response decoding may fail before a middleware even gets a chance to use the right encoding. And it’s not clear how different encoding-declaring elements should be prioritized; in fact, if what @nyov says is true and the current behavior is the same that web browsers follow, the current behavior may be best, and at most we would be looking at implementing a way for users to override that behavior through some kind of setting. This issue may need further discussion.

That said, if whether or not the changes get merged is not relevant for your assignment, feel free to have a go at it. At the very least it will enrich the discussion and possibly help others as a workaround they can implement in their projects.

**Comment by nyov** [comment]
Hello @majate, and thanks for your interest in this! That approach seems absolutely fine and correct.

I would question @Gallaecio if it really makes sense to add a middleware that checks (and potentially rewrites - as in: clones) every response object. That seems like adding more processing overhead that could easily be saved by just implementing it as a potential code branch ("if flag x enabled, then do it that way") in the existing downloader codebase.
Then said code branch would check, and be toggled via, a `Request.meta` flag.

This way it wouldn't be a pluggable middleware, but could be a little cleaner; I think? But I'm not that well-aquainted with the code anymore, and maybe I'm wrong about it.

**Comment by Gallaecio** [comment]
I agree that it may be too much overhead. But since, as you said, response decoding happens at Response class level, also to avoid overhead, it may be best to put the new code somewhere in that part of the code, in the Response class handling of encoding, rather than in a downloader middleware that only gets a response object that has already tried (and possibly failed) decoding the response body.

**Comment by majate** [comment]
Thank you for the information and your ideas! Sadly, we have a quite short time frame for our assignment, so we did not have the possibility to change approach after we had started. As of right now, we have implemented a new downloader middleware that is only included in the middleware pipeline if it is enabled in the settings. The processing of the responses is implemented as:

```python
def process_response(self, request, response, spider):

        if isinstance(response, TextResponse) \
                and response._encoding is None \
                and response._headers_encoding() is not None \
                and response._body_declared_encoding() is not None \
                and response._headers_encoding() != response._body_declared_encoding():
            return response.replace(encoding=response._body_declared_encoding())

        return response
```

This would update the response to obey the encoding defined by the body over the encoding defined in the header, as well as keeping the behaviour of letting an encoding passed in the `__init__` method `encoding` argument remain its priority above both the encoding in the body and the header. Our solution works according to our tests, but there might be some edge cases that we have forgotten to test.

We realise, as you have already mentioned, that it might be better to implement the feature directly in the TextResponse class, instead of needing to process every Response in a middleware. However, this is definitely a quick and easy solution for someone wanting to achieve this behaviour before a better solution is implemented and merged.

Would it be useful for anyone if we created a pull request (just to show our full solution) or would that be unnecessarily?

**Comment by Gallaecio** [comment]
Instead of creating a pull request, I think you could just share the solution here, as people will have to copy-paste it anyway.
